Sometimes runtime error occurs

I am using Ganymede with Docker on my Synology NAS.

Sometimes Ganymede-api throws a runtime error and stops some queues.
When Ganymede-api crashes, Docker restarts it, so Ganymede-api will restart, but queues created before the restart will become dead queues, preventing new queue creation.
Is it possible to solve the above problem?

The log at the time of the error is as follows
```
panic: runtime error: index out of range [0] with length 0
goroutine 2443 [running]:
github.com/zibbp/ganymede/internal/archive.(*Service).TaskVodSaveLiveInfo(0xc0005980f0, 0xc0007edd00, 0xc000851860, 0xc0004dc9a0, 0x1)
	/app/internal/archive/archive.go:605 +0xa69
created by github.com/zibbp/ganymede/internal/archive.(*Service).TaskVodDownloadLiveThumbnail
	/app/internal/archive/archive.go:504 +0x56d

```
Attached is the Docker log
[_ganymede-api_logs.txt](https://github.com/Zibbp/ganymede/files/12113485/_ganymede-api_logs.txt)

@Zibbp 
Hello

The same error as the above user occurred.
The log is as follows, and the part where the actual problem occurs seems to be the same as the user above.
my check interval is 90s.

```
29T04:39:50+09:00 DBG getting live streams using the following query param: ?user_login=sweet_lena0 
panic: runtime error: index out of range [0] with length 0
goroutine 18516 [running]:
github.com/zibbp/ganymede/internal/archive.(*Service).TaskVodSaveLiveInfo(0xc0000a28d0, 0xc000365b00, 0xc0007ac5a0, 0xc0003d89a0, 0x1)
	/app/internal/archive/archive.go:613 +0xab2
created by github.com/zibbp/ganymede/internal/archive.(*Service).TaskVodDownloadLiveThumbnail
	/app/internal/archive/archive.go:546 +0xf9f
usermod: no changes
-------------------------------------
User uid:    1026
User gid:    100
-------------------------------------
Version    : 
2023-09-29T04:40:17+09:00 2023-09-29T04:40:17+09:00 INF 2023-09-29T04:40:17+09:00 2023-09-29T04:40:17+09:00 INF 2023-09-29T04:40:17+09:00 2023-09-29T04:40:17+09:00 INF 2023-09-29T04:40:17+09:00 2023-09-29T04:40:17+09:00 INF config file found at /data/config.json, loading 
Git Hash   : d2941111ae2d3f6c8dbd7572b1bfd958736fed46
Build Time : 2023-09-10T18:34:51Z
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG config file loaded: /data/config.json 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 INF 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 INF 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 INF 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 INF debug mode enabled 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG setting up database connection 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG creating new handler 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG mapping routes 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG setting up twitch auth schedule 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG running twitch auth schedule 
2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG 2023-09-29T04:40:20+09:00 2023-09-29T04:40:20+09:00 DBG authenticating with twitch 
   ____    __
  / __/___/ /  ___
 / _// __/ _ \/ _ \
/___/\__/_//_/\___/ v4.11.1
High performance, minimalist Go web framework
https://echo.labstack.com
____________________________________O/_______
                                    O\
⇨ http server started on [::]:4000
2023-09-29T04:40:21+09:00 2023-09-29T04:40:21+09:00 INF 2023-09-29T04:40:21+09:00 2023-09-29T04:40:21+09:00 INF 2023-09-29T04:40:21+09:00 2023-09-29T04:40:21+09:00 INF 2023-09-29T04:40:21+09:00 2023-09-29T04:40:21+09:00 INF authenticated with twitch 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting up twitch categories schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting up queue item schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting up prune video schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG running set twitch categories schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG running queue item schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF setting up check watched channel videos schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting up check live stream schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting video check interval to run every 180 minutes 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG setting live check interval to run every 90 seconds 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF running check watched channel videos schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG running check live stream schedule 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG checking live channels 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 INF Checking 1 channels for new videos 
2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG 2023-09-29T04:40:25+09:00 2023-09-29T04:40:25+09:00 DBG no queue items are on hold 
```

--- a/internal/archive/archive.go
+++ b/internal/archive/archive.go
@@ -613,9 +613,17 @@ func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue,
 
 	if len(stream.Data) == 0 {
 		log.Error().Msg("stream data is empty")
+		// delete queue item
 		err := database.DB().Client.Queue.DeleteOne(q).Exec(context.Background())
 		if err != nil {
 			log.Error().Err(err).Msg("error deleting queue item")
+		}
+		// delete vod
+		fakeContext := echo.New().NewContext(nil, nil)
+		err = s.VodService.DeleteVod(fakeContext, v.ID, true)
+		if err != nil {
+			log.Error().Err(err).Msg("error deleting vod")
+			return fmt.Errorf("error deleting vod: %v", err)
 		}
 		return fmt.Errorf("stream data is empty")
 	}

--- a/internal/archive/archive.go
+++ b/internal/archive/archive.go
@@ -429,7 +429,14 @@ func (s *Service) RestartTask(c echo.Context, qID uuid.UUID, task string, cont b
 		}
 	case "vod_save_info":
 		if q.LiveArchive {
-			go s.TaskVodSaveLiveInfo(ch, v, q, cont)
+			err = s.TaskVodSaveLiveInfo(ch, v, q, cont)
+			if err != nil {
+				log.Error().Err(err).Msg("error saving live info")
+				q.Update().SetTaskVodSaveInfo(utils.Failed).SaveX(context.Background())
+				s.TaskError(ch, v, q, "vod_save_info")
+				return err
+			}
+			q.Update().SetTaskVodSaveInfo(utils.Success).SaveX(context.Background())
 		} else {
 			go s.TaskVodSaveInfo(ch, v, q, cont)
 		}
@@ -509,7 +516,12 @@ func (s *Service) TaskVodDownloadLiveThumbnail(ch *ent.Channel, v *ent.Vod, q *e
 			// Refresh thumbnails for live stream after 30 minutes
 			go s.RefreshLiveThumbnails(ch, v, q)
 			// Proceed with task
-			go s.TaskVodSaveLiveInfo(ch, v, q, true)
+			err = s.TaskVodSaveLiveInfo(ch, v, q, true)
+			if err != nil {
+				log.Error().Err(err).Msg("error saving live info")
+				q.Update().SetTaskVodSaveInfo(utils.Failed).SaveX(context.Background())
+				s.TaskError(ch, v, q, "vod_save_info")
+			}
 		}
 		return
 	}
@@ -543,7 +555,12 @@ func (s *Service) TaskVodDownloadLiveThumbnail(ch *ent.Channel, v *ent.Vod, q *e
 		// Refresh thumbnails for live stream after 30 minutes
 		go s.RefreshLiveThumbnails(ch, v, q)
 		// Proceed with task
-		go s.TaskVodSaveLiveInfo(ch, v, q, true)
+		err = s.TaskVodSaveLiveInfo(ch, v, q, true)
+		if err != nil {
+			log.Error().Err(err).Msg("error saving live info")
+			q.Update().SetTaskVodSaveInfo(utils.Failed).SaveX(context.Background())
+			s.TaskError(ch, v, q, "vod_save_info")
+		}
 	}
 }
 

--- a/internal/archive/archive.go
+++ b/internal/archive/archive.go
@@ -597,7 +597,7 @@ func (s *Service) TaskVodDownloadThumbnail(ch *ent.Channel, v *ent.Vod, q *ent.Q
 	}
 }
 
-func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue, cont bool) {
+func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue, cont bool) error {
 	log.Debug().Msgf("starting task vod save info for vod %s", v.ID)
 	q.Update().SetTaskVodSaveInfo(utils.Running).SaveX(context.Background())
 
@@ -608,8 +608,18 @@ func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue,
 		log.Error().Err(err).Msg("error fetching twitch vod")
 		q.Update().SetTaskVodSaveInfo(utils.Failed).SaveX(context.Background())
 		s.TaskError(ch, v, q, "vod_save_info")
-		return
+		return err
 	}
+
+	if len(stream.Data) == 0 {
+		log.Error().Msg("stream data is empty")
+		err := database.DB().Client.Queue.DeleteOne(q).Exec(context.Background())
+		if err != nil {
+			log.Error().Err(err).Msg("error deleting queue item")
+		}
+		return fmt.Errorf("stream data is empty")
+	}
+
 	tVod := stream.Data[0]
 
 	err = utils.WriteJson(tVod, fmt.Sprintf("%s/%s", ch.Name, v.FolderName), fmt.Sprintf("%s-info.json", v.FileName))
@@ -617,7 +627,7 @@ func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue,
 		log.Error().Err(err).Msg("error saving info")
 		q.Update().SetTaskVodSaveInfo(utils.Failed).SaveX(context.Background())
 		s.TaskError(ch, v, q, "vod_save_info")
-		return
+		return err
 	}
 	q.Update().SetTaskVodSaveInfo(utils.Success).SaveX(context.Background())
 	if cont {
@@ -632,6 +642,7 @@ func (s *Service) TaskVodSaveLiveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue,
 			go s.TaskLiveChatDownload(ch, v, q, true, busC, startChatDownloadChannel, true)
 		}
 	}
+	return nil
 }
 
 func (s *Service) TaskVodSaveInfo(ch *ent.Channel, v *ent.Vod, q *ent.Queue, cont bool) {

